Extensions/Firebase/B_firebasetools/D_cloudfirestoretools.ts

```diff
@@ -6,6 +6,230 @@ namespace gdjs {
        * @namespace
        */
       export namespace firestore {
+        const queries = new Map<string, firebase.firestore.Query>();
+
+        /**
+         * Converts a firebase document snapshot to a plain dictionary,
+         * so that it may be serialized or converted to a {@link gdjs.Variable}.
+         *
+         * @param doc - The document snapshot.
+         * @returns - The converted object.
+         */
+        const documentSnapshotToSerializable = (
+          doc: firebase.firestore.DocumentSnapshot
+        ) => ({
+          data: doc.data(),
+          exists: doc.exists,
+          id: doc.id,
+        });
+
+        /**
+         * Converts a firebase query snapshot to a plain dictionary,
+         * so that it may be serialized or converted to a {@link gdjs.Variable}.
+         *
+         * @param query - The query snapshot.
+         * @returns - The converted object.
+         */
+        const querySnapshotToSerializable = (
+          query: firebase.firestore.QuerySnapshot
+        ) => ({
+          size: query.size,
+          empty: query.empty,
+          docs: query.docs.map(documentSnapshotToSerializable),
+        });
+
+        /**
+         * Initiate a query over a collection.
+         * @param queryID - The name of the new query.
+         * @param collectionName - The name of the collection to query.
+         */
+        export const startQuery = (queryID: string, collectionName: string) => {
+          queries.set(queryID, firebase.firestore().collection(collectionName));
+        };
+
+        /**
+         * Create a new query from a base query.
+         * @param queryID - The name of the new query.
+         * @param sourceQueryID - The name of the source query.
+         */
+        export const startQueryFrom = (
+          queryID: string,
+          sourceQueryID: string
+        ) => {
+          if (queries.has(sourceQueryID))
+            queries.set(
+              queryID,
+              queries.get(sourceQueryID) as firebase.firestore.Query
+            );
+        };
+
+        /**
+         * Filters out documents whose fields do not match a condition
+         * from a query.
+         * @param queryID - The query to add the filter to.
+         * @param field - The field to run the condition on.
+         * @param op - The condtion operator.
+         * @param value - The value to check against.
+         */
+        export const queryWhere = (
+          queryID: string,
+          field: string,
+          op: Exclude<
+            firebase.firestore.WhereFilterOp,
+            // Exclude unsupported "batch" operations (as they require an array as value to check)
+            'in' | 'array-contains-any' | 'not-in'
+          >,
+          value: string | number
+        ) => {
+          if (queries.has(queryID))
+            queries.set(
+              queryID,
+              (queries.get(queryID) as firebase.firestore.Query).where(
```

Comment:
Strange that you have to do these casts no? 🤔

Reply:
Well, I guess typescript is not smart enough to know that if Map.has is true, Map.get cannot return null

Reply:
Ah right! Note that you could try to use the `!` syntax (`queries.get(queryID)!.where(`) to remove the nullability, if that's the problem (it's not better, it's just shorter). Otherwise not worries.

Reply:
https://github.com/Microsoft/TypeScript/issues/9619